### DISCUSSION TITLE
feat: carapace support

### DISCUSSION THREAD

**Discussion by ShalokShalom** [sammy-ette/Hilbish#305] [Ideas]
Carapace is an universal shell completion library and binary

Reference https://github.com/Rosettea/Hilbish/issues/262#issuecomment-1736409572

**Comment by sammy-ette** [comment]
i tried this a while ago, but i couldn't really figure out a good way to use carapace. if someone else would like to work on it unofficially, please do!

**Reply by jdugan6240** [reply]
Yeah, I looked into this a bit too, but unfortunately carapace doesn't seem to have a good way to generate completions outside of the shells it explicitly supports. I seem to recall there were some circumstances where it _kind of_ works, but it's far from what we would need for a general completer.

It's a shame, because it would be an ideal candidate for this, being exclusively for generating shell completions (I'd like to not have to have the fish shell installed just for good shell completions, personally).

**Reply by jdugan6240** [reply]
Created Carapace issue here: https://github.com/carapace-sh/carapace-bin/issues/2476

**Reply by jdugan6240** [reply]
Ok, according to the linked issue, Carapace has the `export` command which basically does what we need. However, it outputs its completions in json format, not a simple newline-separated series of values. So, while doable, this would probably be best implemented as a plugin rather than a script that can be simply plugged into one's config.

I'll get going on that when I get the chance.

**Reply by sammy-ette** [reply]
> So, while doable, this would probably be best implemented as a plugin rather than a script that can be simply plugged into one's config.

seems plenty easy to do without a plugin. i even have a one liner to decode json in lua.

**Reply by jdugan6240** [reply]
I'd like to see this one liner, because I haven't found a way to parse json to a Lua table without a library (granted, it doesn't need to be a particularly fancy library).

Once you've got a Lua table representing the json, it's just constructing completion groups with the data.

**Reply by sammy-ette** [reply]
```lua
load('return ' .. jsonstr:gsub('("[^"]-"):', '[%1]='))()
```

i don't remember where i got this from but i used it since i didn't want to get the rxi json library just for my image upload script [(which, yes, is in lua)](https://github.com/TorchedSammy/dotfiles/blob/master/bin/ssf)

**Reply by jdugan6240** [reply]
Unfortunately, that breaks when string values in the json table happen to contain a `":`, as in the following example:

```lua
jsonstr = '{"test": "hello \\":"}'
obj = load('return ' .. jsonstr:gsub('("[^"]-"):', '[%1]='))()
```

The above example crashes because the resulting string that's passed to `load` isn't valid Lua. The rxi json library of course handles this just fine.

Is it _likely_ that carapace completion output would feature this edge case? No, but it is definitely possible.

So, as clever as that one liner is, I don't think we can rely on it for a robust carapace script. Plus, `json.decode(jsonstr)` is a lot easier to understand at a glance.
